src/Data/Nat/Primality.agda

```diff
@@ -8,80 +8,290 @@
+0-rough : Rough 0 n
+0-rough (boundedComposite () _)
+
+1-rough : Rough 1 n
+1-rough (boundedComposite ⦃()⦄ z<s _)
+
+2-rough : Rough 2 n
+2-rough (boundedComposite ⦃()⦄ (s<s z<s) _)
 
--- Definition of primality.
+-- If a number n > 1 is k-rough, then k ≤ n
+rough⇒≤ : ⦃ NonTrivial n ⦄ → Rough k n → k ≤ n
+rough⇒≤ rough = ≮⇒≥ λ k>n → rough (boundedComposite k>n ∣-refl)
+
+-- If a number n is k-rough, and k ∤ n, then n is (suc k)-rough
+∤⇒rough-suc : k ∤ n → Rough k n → Rough (suc k) n
+∤⇒rough-suc k∤n r (boundedComposite d<1+k d∣n) with m<1+n⇒m<n∨m≡n d<1+k
+... | inj₁ d<k      = r (boundedComposite d<k d∣n)
+... | inj₂ d≡k@refl = contradiction d∣n k∤n
+
+-- If a number is k-rough, then so are all of its divisors
+rough⇒∣⇒rough : Rough k m → n ∣ m → Rough k n
+rough⇒∣⇒rough r n∣m (boundedComposite d<k d∣n)
+  = r (boundedComposite d<k (∣-trans d∣n n∣m))
+
+------------------------------------------------------------------------
+-- Corollary: relationship between roughness and primality
+
+-- If a number n is p-rough, and p > 1 divides n, then p must be prime
+rough⇒∣⇒prime : ⦃ NonTrivial p ⦄ → Rough p n → p ∣ n → Prime p
+rough⇒∣⇒prime r p∣n = prime (rough⇒∣⇒rough r p∣n)
+
+------------------------------------------------------------------------
+-- Basic (non-)instances of Composite
+
+¬composite[0] : ¬ Composite 0
+¬composite[0] (_ , composite[0]) = 0-rough composite[0]
+
+¬composite[1] : ¬ Composite 1
+¬composite[1] (_ , composite[1]) = 1-rough composite[1]
+
+composite[4] : Composite 4
+composite[4] = composite {d = 2} (λ ()) (divides-refl 2)
+
+------------------------------------------------------------------------
+-- Basic (non-)instances of Prime
+
+¬prime[0] : ¬ Prime 0
+¬prime[0] ()
+
+¬prime[1] : ¬ Prime 1
+¬prime[1] ()
+
+prime[2] : Prime 2
+prime[2] = prime 2-rough
+
+------------------------------------------------------------------------
+-- Basic (non-)instances of Irreducible
+
+¬irreducible[0] : ¬ Irreducible 0
+¬irreducible[0] irr[0] = [ (λ ()) , (λ ()) ]′ (irr[0] {2} (divides-refl 0))
+
+irreducible[1] : Irreducible 1
+irreducible[1] m|1 = inj₁ (∣1⇒≡1 m|1)
+
+irreducible[2] : Irreducible 2
+irreducible[2] {zero}  0∣2 with () ← 0∣⇒≡0 0∣2
+irreducible[2] {suc _} d∣2 with ∣⇒≤ d∣2
+... | z<s     = inj₁ refl
+... | s<s z<s = inj₂ refl
+
+
+------------------------------------------------------------------------
+-- NonTrivial
+
+composite⇒nonTrivial : Composite n → NonTrivial n
+composite⇒nonTrivial {1}      = flip contradiction ¬composite[1]
+composite⇒nonTrivial {2+ _} _ = _
+
+prime⇒nonTrivial : Prime p → NonTrivial p
+prime⇒nonTrivial (prime ⦃ nt ⦄ _) = nt
+
+------------------------------------------------------------------------
+-- NonZero
+
+composite⇒nonZero : Composite n → NonZero n
+composite⇒nonZero {suc _} _ = _
+
+prime⇒nonZero : Prime p → NonZero p
+prime⇒nonZero (prime _) = nonTrivial⇒nonZero
+
+irreducible⇒nonZero : Irreducible n → NonZero n
+irreducible⇒nonZero {zero} = flip contradiction ¬irreducible[0]
```

Comment:
Same comment as `composite⇒nonTrivial`

Reply:
Yes... but the typechecker gets fouled up in the implicit vs. explicit quantification, and I just don't have the patience to figure those things out!